Adopt more smart pointers in dom/Text

jelee53 opened the pull request:
#### dcd17064b3b5f57c9b57b09f5a061c0560ac9e32
<pre>
Adopt more smart pointers in dom/Text
<a href="https://bugs.webkit.org/show_bug.cgi?id=294640">https://bugs.webkit.org/show_bug.cgi?id=294640</a>
<a href="https://rdar.apple.com/153684504">rdar://153684504</a> (Adopt more smart pointers in dom/Text (294640))

Reviewed by Ryosuke Niwa.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::wholeText const):
* Source/WebCore/dom/TextNodeTraversal.cpp:
(WebCore::TextNodeTraversal::appendContents):
(WebCore::TextNodeTraversal::childTextContent):
* Source/WebCore/dom/TextNodeTraversal.h:
(WebCore::TextNodeTraversal::firstTextWithinTemplate):
(WebCore::TextNodeTraversal::traverseNextTextTemplate):

Canonical link: <a href="https://commits.webkit.org/296685@main">https://commits.webkit.org/296685@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20a3947b940c8130fbcb784b92b2f6feb8ee1d9a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108990 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28650 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19075 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114200 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59317 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110953 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29333 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37217 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82825 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111938 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23324 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98171 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63267 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22738 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16309 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58895 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92695 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16356 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117319 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36039 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26637 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91841 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94435 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91642 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36552 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14307 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31900 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17631 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35936 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41454 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35635 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38975 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/af255e6c81949a2f6691d71d0f06e863a87fd4a6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46876

jelee53 commented on Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
```diff
@@ -24290,6 +24290,9 @@
 		FB39D0701200ED9200088E69 /* Project object */ = {
 			isa = PBXProject;
 			attributes = {
+				KnownAssetTags = (
```
Trying to figure out how to force push a new commit that overwrites this... 

Please ignore this for now.

jelee53 replied:
Done. Updated commit should remove this.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d0244c2abd6944e24d2d5feb30f5ef7fca34f140. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46876

RupinMittal commented on Source/WebCore/dom/Text.cpp:
```diff
@@ -104,13 +104,12 @@ static const Text* latestLogicallyAdjacentTextNode(const Text* text)
-    const Text* endText = latestLogicallyAdjacentTextNode(this);
-    ASSERT(endText);
-    const Node* onePastEndText = TextNodeTraversal::nextSibling(*endText);
+    RefPtr startText = const_cast<Text*>(earliestLogicallyAdjacentTextNode(this));
```
I believe we don't need the `const_cast<Text*>` here. You should be able to just do:

`RefPtr startText = earliestLogicallyAdjacentTextNode(this);`

Same goes for the other uses of `const_cast` here--could remove all of them.

RupinMittal commented on Source/WebCore/dom/Text.cpp:
```diff
@@ -104,13 +104,12 @@ static const Text* latestLogicallyAdjacentTextNode(const Text* text)
 {
-    const Text* startText = earliestLogicallyAdjacentTextNode(this);
-    const Text* endText = latestLogicallyAdjacentTextNode(this);
-    ASSERT(endText);
```
We should probably keep this assert.

RupinMittal reviewed (commented):
Looks mostly good, a few small comments.

RupinMittal commented on Source/WebCore/dom/Text.cpp:
```diff
@@ -104,13 +104,12 @@ static const Text* latestLogicallyAdjacentTextNode(const Text* text)
-    ASSERT(endText);
-    const Node* onePastEndText = TextNodeTraversal::nextSibling(*endText);
+    RefPtr startText = const_cast<Text*>(earliestLogicallyAdjacentTextNode(this));
+    RefPtr endText = const_cast<Text*>(latestLogicallyAdjacentTextNode(this));
```
Same here.

jelee53 replied:
Done! Thanks for pointing this out Rupin!

RupinMittal commented on Source/WebCore/dom/Text.cpp:
```diff
@@ -104,13 +104,12 @@ static const Text* latestLogicallyAdjacentTextNode(const Text* text)
-    const Node* onePastEndText = TextNodeTraversal::nextSibling(*endText);
+    RefPtr startText = const_cast<Text*>(earliestLogicallyAdjacentTextNode(this));
+    RefPtr endText = const_cast<Text*>(latestLogicallyAdjacentTextNode(this));
+    RefPtr onePastEndText = const_cast<Text*>(TextNodeTraversal::nextSibling(*endText));
```
Maybe use:

`RefPtr<const Text> onePastEndText = TextNodeTraversal::nextSibling(*endText);`

RupinMittal commented on Source/WebCore/dom/Text.cpp:
```diff
@@ -104,13 +104,12 @@ static const Text* latestLogicallyAdjacentTextNode(const Text* text)
 
     StringBuilder result;
-    for (const Text* text = startText; text != onePastEndText; text = TextNodeTraversal::nextSibling(*text))
+    for (RefPtr text = startText; text != onePastEndText; text = const_cast<Text*>(TextNodeTraversal::nextSibling(*text)))
```
Same as above.

cdumez commented:
> Adopt more smart pointers in dom/Text
> https://bugs.webkit.org/show_bug.cgi?id=294640
> Include a Radar link (OOPS!).

Please drop the OOPS line or put the radar link if you have one (not mandatory).

cdumez commented:
When starting out, I'd recommend staying out of dom/ as this is performance sensitive code. This PR will have to be A/B tested on Speedometer.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8c691e4b8cc9416b3c651b4f84630fce7294ad80. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46876

RupinMittal commented on Source/WebCore/dom/Text.cpp:
```diff
@@ -104,13 +104,13 @@ static const Text* latestLogicallyAdjacentTextNode(const Text* text)
+    RefPtr endText = latestLogicallyAdjacentTextNode(this);
     ASSERT(endText);
-    const Node* onePastEndText = TextNodeTraversal::nextSibling(*endText);
+    RefPtr<const Text> onePastEndText = TextNodeTraversal::nextSibling(*endText);
```
Let's keep using `Node` here. So `RefPtr<const Node>`.

jelee53 replied:
Oops I missed that it was using Node originally. Thanks for catching that Rupin!

RupinMittal commented on Source/WebCore/dom/Text.cpp:
```diff
@@ -104,13 +104,13 @@ static const Text* latestLogicallyAdjacentTextNode(const Text* text)
 
     StringBuilder result;
-    for (const Text* text = startText; text != onePastEndText; text = TextNodeTraversal::nextSibling(*text))
+    for (RefPtr<const Text> text = startText; text != onePastEndText; text = TextNodeTraversal::nextSibling(*text))
```
Since `startText` is `RefPtr<const Text>` from above, we don't need to specify the `<const Text>` part again. We can just use `RefPtr text` here.

jelee53 replied:
Ah I see. When we create `RefPtr text` variable., `RefPtr` is automatically created with the same type as `StartText` (aka a constText as the typeParameter). Good point! Will fix this - thanks Rupin!

RupinMittal reviewed (commented):
Almost there, but as Chris mentioned, this will have to be A/B tested before landing.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/20a3947b940c8130fbcb784b92b2f6feb8ee1d9a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46876

jelee53 committed:
Adopt more smart pointers in dom/Text
https://bugs.webkit.org/show_bug.cgi?id=294640
rdar://153684504 (Adopt more smart pointers in dom/Text (294640))

Reviewed by Ryosuke Niwa.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::wholeText const):
* Source/WebCore/dom/TextNodeTraversal.cpp:
(WebCore::TextNodeTraversal::appendContents):
(WebCore::TextNodeTraversal::childTextContent):
* Source/WebCore/dom/TextNodeTraversal.h:
(WebCore::TextNodeTraversal::firstTextWithinTemplate):
(WebCore::TextNodeTraversal::traverseNextTextTemplate):

Canonical link: https://commits.webkit.org/296685@main
